### PR TITLE
Clusters Sidenav Clusterly Changes 

### PR DESCRIPTION
**Pull request by chenraff**
* Improve queries' dialog visibility + add an option to select all queries or delete queries through the dialog.
* Update Clusterly back-end & front-end to support information needed for the clusters' sidenav.
* Add delete (X) and add similar (+) buttons to each clusters that appears on hover.
* Add 'add-similar-dialog' that allows the user to control which similar cluster they want to add to the screen.

### CODE REVIEW

File: trendly/src/app/queries-dialog/queries-dialog.component.html [hunk 1]
```diff
@@ -1,21 +1,51 @@
...
+      <mat-label>Select Location:</mat-label>
+      <mat-select (selectionChange)="selectedCluster = $event.value">
+        <mat-option *ngFor="let cluster of data.clusters"
+                    [value]="cluster">{{cluster.title}}</mat-option>
       </mat-select>
     </mat-form-field>
   </div>
   <div class="btn">
-    <button class='apply-button' (click)="data.updateFunc(selectedCluster, terms.selectedOptions.selected, data.currentCluster, data.clusterly)">Apply</button>
+    <button class='apply-button'
```
Comment (adiassulin) [inline trendly/src/app/queries-dialog/queries-dialog.component.html:48]: "  " instead of ' '

File: trendly/src/backend/QueriesExpansion.java [hunk 2]
```diff
@@ -134,14 +131,24 @@
...
+  private static double calculateClusterVolume(TrendsQuery[] queries) {
+    return Arrays.stream(queries).mapToDouble((query) -> query.value).sum();
+  }
+
+  /** Returns new cluster object based on the giben queries, id and title. */
+  private static Cluster createNewCluster(TrendsQuery[] queries, AtomicInteger id, String title) {
+    double volume = calculateClusterVolume(queries);
+    TrendsQuery[] queriesToDisplay =
+        Arrays.copyOfRange(queries, 0, Math.min(MAX_QUERIES, queries.length));
+    TrendsQuery[] additionalQueries =
```
Comment (mayascha) [inline trendly/src/backend/QueriesExpansion.java:150]: better to caculate Math.min(MAX_QUERIES, queries.length) once and use in both

**Review by mayascha — APPROVED** [review]

File: trendly/src/app/clusters-section/clusters-section.component.ts [hunk 3]
```diff
@@ -257,21 +331,126 @@ export class ClustersSectionComponent {
...
-            .attr('class', 'clusters-titles')
-            .text(this.clusters.get(clusterID).title)
-            .attr('x', location.xPosition - window.innerWidth / 80)
-            .attr(
-                'y',
-                location.yPosition - Math.min(window.innerHeight / 4, 200));
+        const x: number = location.xPosition - window.innerWidth / 80;
+        const y: number =
+            location.yPosition - Math.min(window.innerHeight / 4, 200);
+        const nonDisplayedSimCluster = this.getNonDisplayedSimilar(clusterID);
```
Comment (adiassulin) [inline trendly/src/app/clusters-section/clusters-section.component.ts:341]: add type number[] for nonDisplayedSimCluster variable

File: trendly/src/app/clusters-section/clusters-section.component.ts [hunk 4]
```diff
@@ -257,21 +331,126 @@ export class ClustersSectionComponent {
...
+        .attr('y', y);
+  }
+
+  /**
+   * Adds to titleBtnGroup the add similar clusters button (when clicked on adds
+   * to the displayed clusters the related clusters ids).
+   */
+  private addSimilarBtn(
+      titleBtnGroup, x: number, y: number, clusterID: number,
+      nonDisplayedSimCluster: number[]) {
```
Comment (adiassulin) [inline trendly/src/app/clusters-section/clusters-section.component.ts:377]: add function return value type - void (also at the next one)

File: trendly/src/app/clusters-section/clusters-section.component.ts [hunk 5]
```diff
@@ -449,9 +633,14 @@ export class ClustersSectionComponent {
...
   private addCluster(title: string, clusterly: ClustersSectionComponent): void {
     const newCluster: Cluster =
-        new Cluster(title, clusterly.clusters.size + 1, []);
+        new Cluster(title, clusterly.clusters.size + 1, 0, [], [], []);
+    clusterly.clustersToDisplay.set(clusterly.clusters.size + 1, newCluster);
     clusterly.clusters.set(clusterly.clusters.size + 1, newCluster);
-    clusterly.simulation.stop();
+    clusterly.displayedClustersDiffer.set(
+        newCluster.id, clusterly.differs.find(newCluster).create());
+    if (clusterly.simulation) {
```
Comment (adiassulin) [inline trendly/src/app/clusters-section/clusters-section.component.ts:641]: maybe with ternary operator

File: trendly/src/app/add-similar-dialog/add-similar-dialog.component.ts [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+  }
+
+  /**
+   * Selects all rows if they are not all selected; otherwise clear selection.
+   */
+  masterToggle(): void {
+    this.isAllSelected() ?
+        this.selectedClusters.clear() :
+        this.dataSource.data.forEach(
+            cluster => {this.selectedClusters.select(cluster)});
```
Comment (mayascha) [inline trendly/src/app/add-similar-dialog/add-similar-dialog.component.ts:53]: I don't think you need the {}

**Review by mayascha — APPROVED** [review]
Great work!

**Pull request merged**